core/utils/historied-data/src/linear.rs

```diff
@@ -0,0 +1,488 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.	If not, see <http://www.gnu.org/licenses/>.
+
+//! Transactional overlay implementation.
+//!
+//! This follows a linear succession of states.
+//! This contains multiple unbounded transaction layer
+//! and an additional top level 'prospective' layer.
+//! It only allows linear history (no branch so
+//! inner storage is only an array of element).
+
+use crate::State as TransactionState;
+use rstd::vec::Vec;
+use rstd::vec;
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(not(feature = "std"))]
+type MemoryOnly<V> = Vec<(V, usize)>;
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(feature = "std")]
+type MemoryOnly<V> = smallvec::SmallVec<[(V, usize); ALLOCATED_HISTORY]>;
+
+/// Size of preallocated history per element.
+/// Currently at two for committed and prospective only.
+/// It means that using transaction in a module got a direct allocation cost.
+#[cfg(feature = "std")]
+const ALLOCATED_HISTORY: usize = 2;
+
+/// History of value that are related to a state history (eg field `history` of
+/// an `OverlayedChangeSet`).
+///
+/// Values are always paired with a state history index.
+#[derive(Debug, Clone)]
+#[cfg_attr(any(test, feature = "test"), derive(PartialEq))]
+pub struct History<V>(MemoryOnly<V>);
+
+impl<V> Default for History<V> {
+	fn default() -> Self {
+		History(Default::default())
+	}
+}
+
+// Following implementation are here to isolate
+// buffer specific functions.
+impl<V> History<V> {
+
+	fn get_state(&self, index: usize) -> (&V, usize) {
+		(&self.0[index].0, self.0[index].1)
+	}
+
+	#[cfg(any(test, feature = "test"))]
+	/// Create an history from an existing history.
+	pub fn from_iter(input: impl IntoIterator<Item = (V, usize)>) -> Self {
+		let mut history = History::default();
+		for v in input {
+			history.unsafe_push(v.0, v.1);
+		}
+		history
+	}
+
+	#[cfg(any(test, feature = "test"))]
+	/// Debugging function for test and fuzzing.
+	pub fn internal_item_counts(&self) -> usize {
+		self.0.len()
+	}
```

Comment:
it is just here to expose `len` as public (mainly to test gc).

Reply:
yeah, but I mean look at the [`Self::len`](https://github.com/paritytech/substrate/pull/3263/files/b534b3d2a070ba7fd7c264ff042f4a20b0ba92bd#diff-e6eb5324fa37219c00c89cbfb41558bcR85), it has the same implementation basically. Why don't just call it instead of `internal_item_counts`.

Reply:
I can do like that https://github.com/paritytech/substrate/pull/3263/commits/d9d21185c012f14687d7a322fda1d9bced8af570 , but I like `internal_item_counts` name better.